crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -896,11 +934,29 @@ impl Object {
                     checkpoint_sequence_number: history_object.checkpoint_sequence_number,
                 }),
                 checkpoint_viewed_at,
+                root_version: None,
             }),
         }
     }
 }
 
+/// We're deliberately choosing to use a child object's version as the root here, and letting the
+/// caller override it with the actual root object's version if it has access to it.
+///
+/// Using the child object's version as the root means that we're seeing the dynamic field tree
+/// under this object at the state resulting from the transaction that produced this version.
+///
+/// See [`Object::root_version`] for more details on parent/child object version mechanics.
+fn version_for_dynamic_fields(native: &NativeObject) -> Option<u64> {
+    use NativeOwner::*;
+    match native.as_inner().owner {
+        AddressOwner(_) | Shared { .. } | ObjectOwner(_) => {
+            Some(native.as_inner().version().into())
+        }
+        Immutable => None,
```

Comment:
Is there any reason why we can't use `Some(native.version())` instead? I suppose because `Immutable` freezes the object version so it's no longer useful as a means to give us greater historical granularity (compared to just using `checkpoint_sequence_number`)

Reply:
Yeah, that can be simpler. It should work assuming:
* the 'freeze' operation bumps the version of the parent object; it should since we're acquiring it by value? (write lock)
* removing a DOF from its parent and freezing it (if that is possible) always bumps its version

That last point is important since we may have the "Parent(6) -> Child1 (5) -> Child2 (6)" scenario. What happens to Child1's version if we remove it from Parent and freeze it without changing its contents?

Reply:
Otherwise we have this scenario where querying Child1 (5) will not return Child1 (6) and the former has owner `Immutable` so we don't have a direct path to its previous root's version (at the time Child1 was frozen) that we can use in the upcoming `Query.owner(address, rootVersion)`